Export `DecompressError` + Improve Documentation

Anteproperispomenon opened the pull request:
Functionally, this just adds `DecompressError(..)` to the export list of `Codec.Compression.BZip.Internal`, but it also adds
a lot of documentation to the `Internal` module from the corresponding functions from [zlib](https://hackage.haskell.org/package/zlib). I also added some documentation for the arguments of the `foldCompressStream`/etc... functions to improve clarity.

Also, I bumped the version number and updated the changelog, which I suspect I wasn't supposed to? This is my first time contributing upstream to a Haskell package, so I wasn't sure of the protocol.

Anteproperispomenon committed:
Add `DecompressError` to export list.

It was missing from Codec.Compression.BZip.Internal.

Anteproperispomenon committed:
Improved documentation of Internal.hs

Copied (and fixed) documentation from `zlib`, and also added some specific documentation to the fold functions.

Also bumped version number and added details to changelog.

Bodigrim commented on CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.5.2.1
```
If there is an API change (such as a new export), it should be `0.5.3.0`. The fourth digit is just for packaging changes, see https://pvp.haskell.org for details.

Anteproperispomenon replied:
Done

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -284,7 +401,11 @@ compressStream (CompressParams blockSize workFactor initChunkSize) =
+-- of compressed data chunks as input and receive a sequence of uncompressed
+-- data chunks as output. The process is incremental, in that the demand for
+-- input and provision of output are interleaved.
+--
```
Since this was not exported previously, could you please add `@since@ annotation?

Anteproperispomenon replied:
Done!

Bodigrim reviewed (commented):
Thanks!

Anteproperispomenon committed:
Improved Documentation + Changed Version

Fixed documentation of function arguments to use `-- ^` instead of `-- |`, since it should work with more versions of Haddock. Also added some examples to the `fold` argument of foldCompressStream/foldDecompressStream.

Changed version number to be 0.5.3.0 instead of 0.5.2.1, since the change introduces a new export.

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -145,11 +221,34 @@ data CompressStream m =
+  -- second argument is already embedded in the
+  -- monad. This is typically a lambda of the form
+  -- 
+  -- > \chunk next -> L.chunk chunk <$> next
```
A reader does not necessarily know at this point that `L.chunk` is `Data.ByteString.Lazy.Internal.chunk`. I'd suggest to spell it explicitly.

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -145,11 +221,34 @@ data CompressStream m =
+  -> m a
+  -- ^ The base value of the fold. If the output
+  -- is itself a 'L.ByteString', this can just
+  -- be (return 'L.empty').
```
```suggestion
  -- be (@return@ 'L.empty').
```

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -145,11 +221,34 @@ data CompressStream m =
+  -- be (return 'L.empty').
+  -> CompressStream m 
+  -- ^ The input stream. Typically, this is
+  -- ('compressIO' params) or ('compressST' params),
```
```suggestion
  -- ('compressIO' @params@) or ('compressST' @params@),
```

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -170,11 +269,19 @@ foldCompressStream input output end = fold
+  -- this is just 'L.empty' or 'mempty'.
+  -> (forall s. CompressStream (ST s))
+  -- ^ The compression stream. Typically, this is
+  -- ('compressST' params).
```
```suggestion
  -- ('compressST' @params@).
```

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -319,12 +456,44 @@ modprefix = ("Codec.Compression.BZip: " ++)
+  -- second argument is already embedded in the
+  -- monad. This is typically a lambda of the form
+  -- 
+  -- > \chunk next -> L.chunk chunk <$> next
```
Same as above, let's use a fully-qualified name.

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -319,12 +456,44 @@ modprefix = ("Codec.Compression.BZip: " ++)
+  -- is the base value of the right-fold operation.
+  -> (DecompressError -> m a)
+  -- ^ How to handle errors. Typically, this is
+  -- 'throw', but it can be e.g. (return . 'Left')
```
```suggestion
  -- 'throw', but it can be e.g. (@return@ . 'Left')
```

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -319,12 +456,44 @@ modprefix = ("Codec.Compression.BZip: " ++)
+  -- if the output value is wrapped in 'Either'.
+  -> DecompressStream m 
+  -- ^ The input stream. Typically, this is
+  -- ('decompressIO' params) or ('decompressST' params),
```
```suggestion
  -- ('decompressIO' @params@) or ('decompressST' @params@),
```

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -336,12 +505,38 @@ foldDecompressStream input output end err = fold
+  -- error, just use 'throw'.
+  -> (forall s. DecompressStream (ST s))
+  -- ^ The decompression stream. Typically, this is
+  -- ('decompressST' params).  
```
```suggestion
  -- ('decompressST' @params@).  
```

Bodigrim commented on Codec/Compression/BZip/Internal.hs:
```diff
@@ -361,8 +556,23 @@ foldDecompressStreamWithInput chunk end err = \s lbs ->
+-- | Decompress a data stream provided as a lazy 'L.ByteString'.
+--
+-- It will throw an exception if any error is encountered in the input data.
+-- If you need more control over error handling then use one the incremental
```
```suggestion
-- If you need more control over error handling then use one of the incremental
```

Anteproperispomenon committed:
Fixed documentation issues.

Bodigrim commented:
Thanks, awesome! Do you need a release?
